nifi-nar-bundles/nifi-flow-registry-client-bundle/nifi-flow-registry-client-services/src/main/java/org/apache/nifi/registry/flow/NifiRegistryFlowRegistryClient.java

```diff
@@ -0,0 +1,376 @@
+            .build();
+    public final static PropertyDescriptor PROPERTY_TRUSTSTORE_PASSWORD = new PropertyDescriptor.Builder()
+            .name("truststorePassword")
+            .displayName("Truststore Password")
+            .description("The password for the Truststore")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .sensitive(true)
+            .required(false)
+            .build();
+    public final static PropertyDescriptor PROPERTY_TRUSTSTORE_TYPE = new PropertyDescriptor.Builder()
+            .name("truststoreType")
+            .displayName("Truststore Type")
+            .description("The Type of the Truststore")
+            .allowableValues(KeystoreType.values())
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .required(false)
+            .build();
+
+    private volatile String registryUrl;
+    private volatile NiFiRegistryClient registryClient;
+
+    private synchronized NiFiRegistryClient getRegistryClient(final FlowRegistryClientConfigurationContext context) {
+        final String configuredUrl = context.getProperty(PROPERTY_URL).evaluateAttributeExpressions().getValue();
+
+        final URI uri;
+
+        try {
+            // Handles case where the URI entered has a trailing slash, or includes the trailing /nifi-registry-api
+            uri = new URIBuilder(configuredUrl).setPath("").removeQuery().build();
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException("The given Registry URL is not valid: " + configuredUrl);
+        }
+
+        final String uriScheme = uri.getScheme();
+        if (uriScheme == null) {
+            throw new IllegalArgumentException("The given Registry URL is not valid: " + configuredUrl);
+        }
+
+        final String proposedUrl = uri.toString();;
+
+        if (!proposedUrl.equals(registryUrl)) {
+            registryUrl = proposedUrl;
+            invalidateClient();
+        }
+
+        if (registryClient != null) {
+            return registryClient;
+        }
+
+        final NiFiRegistryClientConfig config = new NiFiRegistryClientConfig.Builder()
+                .connectTimeout(30000)
+                .readTimeout(30000)
+                .sslContext(extractSSLContext(context))
+                .baseUrl(registryUrl)
+                .build();
+        registryClient = new JerseyNiFiRegistryClient.Builder()
+                .config(config)
+                .build();
+
+        return registryClient;
+    }
+
+    @Override
+    protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return Arrays.asList(
+                PROPERTY_URL,
+                PROPERTY_KEYSTORE_PATH,
+                PROPERTY_KEYSTORE_TYPE,
+                PROPERTY_KEYSTORE_PASSWORD,
+                PROPERTY_KEY_PASSWORD,
+                PROPERTY_TRUSTSTORE_PATH,
+                PROPERTY_TRUSTSTORE_TYPE,
+                PROPERTY_TRUSTSTORE_PASSWORD
```

Comment:
We shouldn't need keystore/truststore properties. Instead we should just take an SSLContext Service.

Reply:
Yeah, I was considering that but as registries are not part of a process group (not even the root) contrary to the controller services, this I believe would bring in some issues